pkg/cluster/contour_certs.go

```diff
@@ -0,0 +1,174 @@
+package cluster
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/projectcontour/contour/pkg/certs"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/client-go/util/cert"
+)
+
+func (c *Controller) RotateContourCerts() error {
+	contourNamespace := c.Config.ContourCertNamespace
+	contourSecretName := c.Config.ContourCertSecret
+	envoySecretName := c.Config.EnvoyCertSecret
+
+	if contourNamespace == "" || contourSecretName == "" || envoySecretName == "" {
+		c.Log.Debugf("Contour namespace and secrets not set, skipping certificates renewal")
+		return nil
+	}
+
+	caCert, contourCert, envoyCert, err := c.readContourCerts(contourNamespace, contourSecretName, envoySecretName)
+	if err != nil {
+		return errors.Wrapf(err, "read contour certificates")
+	}
+	if caCert == nil {
+		c.Log.Debugf("Contour ca cert not found, skipping renewal")
+		return nil
+	}
+	if contourCert == nil {
+		c.Log.Debugf("Contour cert not found, skipping renewal")
+		return nil
+	}
+	if envoyCert == nil {
+		c.Log.Debugf("Envoy cert not found, skipping renewal")
+		return nil
+	}
+
+	if !c.shouldRotateContourCerts(caCert, contourCert, envoyCert) {
+		c.Log.Debugf("Contour certs have more than %s until expiration, skipping renewal", duration.ShortHumanDuration(c.Config.RotateCertsTTL))
+		return nil
+	}
+
+	if err := c.updateContourCerts(contourNamespace, contourSecretName, envoySecretName); err != nil {
+		return errors.Wrap(err, "update certs")
+	}
+
+	c.Log.Infof("Renewed contour and envoy certs")
+
+	selector := metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": "envoy"}).String(),
+	}
+	if err := c.Config.Client.CoreV1().Pods(contourNamespace).DeleteCollection(context.TODO(), metav1.DeleteOptions{}, selector); err != nil {
```

Comment:
This would cause brief downtime. I wonder if there is a better way to rollout the change.

Reply:
yeah, i thought about that too, not sure if there's a better way

Reply:
the official documentation for contour says that the envoy pods need to be restarted